src/Orleans.Streaming.RabbitMQ/Providers/Streams/RabbitMQ/RabbitMQAdapterReceiver.cs

```diff
@@ -0,0 +1,101 @@
+
+        public static IQueueAdapterReceiver Create(SerializationManager serializationManager,
+                                                   ILoggerFactory loggerFactory,
+                                                   QueueId queueId,
+                                                   IRabbitMQDataAdapter dataAdapter,
+                                                   RabbitMQOptions options)
+        {
+            if (queueId == null) throw new ArgumentNullException(nameof(queueId));
+            if (dataAdapter == null) throw new ArgumentNullException(nameof(dataAdapter));
+            if (serializationManager == null) throw new ArgumentNullException(nameof(serializationManager));
+
+            return new RabbitMQAdapterReceiver(options, loggerFactory, queueId, serializationManager, dataAdapter);
+        }
+
+        private RabbitMQAdapterReceiver(RabbitMQOptions options,
+                                        ILoggerFactory loggerFactory,
+                                        QueueId queueId,
+                                        SerializationManager serializationManager,
+                                        IRabbitMQDataAdapter rabbitMQAdapter)
+        {
+            _rabbitFact = options ?? throw new ArgumentNullException(nameof(options));
+            _logger = loggerFactory.CreateLogger($"{this.GetType().FullName}.{queueId}");
+            _serializationManager = serializationManager;
+            _dataAdapter = rabbitMQAdapter;
+        }
+
+        public Task Initialize(TimeSpan timeout)
+        {
+            _manager = new RabbitMQManager(_rabbitFact, _logger);
+            if (_logger.IsEnabled(LogLevel.Trace))
+            {
+                _logger.LogTrace("instantiated rabbitmq colony");
+            }
+            return Task.CompletedTask;
+        }
+
+        public async Task Shutdown(TimeSpan timeout)
+        {
+            try
+            {
+                if (_outstandingTask != null)
+                {
+                    await _outstandingTask;
+                }
+            }
+            finally
+            {
+                _manager.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Retrieves a message from the queue.
+        /// </summary>
+        /// <param name="maxCount">Number of messages to receive.</param>
+        /// <returns></returns>
+        public Task<IList<IBatchContainer>> GetQueueMessagesAsync(int maxCount = 1)
+        {
+            var colonyRef = _manager;
+            if (colonyRef == null) return Task.FromResult<IList<IBatchContainer>>(null);
+
+            var upstream = new List<IBatchContainer>();
+
+            // rabbitmq does not have a pullN interface, so we just loop through how many messages we want to pull.
+            for (var i = 0; i < maxCount; i++)
+            {
+                var msg = colonyRef.ReceiveMessage();
+                var container = _dataAdapter.FromPullResponseMessage(msg, _lastReadMessage++);
```

Comment:
msg will be `null` if there are no messages in queue (or when `maxCount` is greater than rabbitmq's Prefetch Limit).

Reply:
I currently have the logic set to acknowledge every message received, so there shouldn't be too much concern on hitting the prefetch limit (at least based on my understanding of prefetching in Rabbit) - I can definitely document that when I figure out where to put it. It's conservative, but it's safer for clients, and the `basic.qos` tag can be set on the RabbitMQ side to adjust as needed.